src/map/unit.c

```diff
@@ -1155,7 +1155,7 @@ static int unit_skilluse_id(struct block_list *src, int target_id, uint16 skill_
 	int ret = unit->skilluse_id2(src, target_id, skill_id, skill_lv, casttime, castcancel);
 	struct map_session_data *sd = BL_CAST(BL_PC, src);
 
-	if (sd != NULL)
+	if (sd != NULL && ret == 0)
```

Comment:
this mean after this change auto cast never cleaned if skill was used without errors?
i think this wrong

Reply:
`unit->skilluse_id2()` returns `0` on failure, so this is okay.

Maybe the
```c
else if (sd != NULL && ret != 0 && skill_id != SA_ABRACADABRA && skill_id != WM_RANDOMIZESPELL)
		skill->validate_autocast_data(sd, skill_id, skill_lv);
```
block I removed in #2699 [(LINK)](https://github.com/HerculesWS/Hercules/pull/2699/files?file-filters%5B%5D=.c&file-filters%5B%5D=.inc#diff-fc6e84e111da95e50d06e579cf274053) should be re-added, too.
I currently can't have a closer look at this, so this is just a suggestion...

Reply:
@4144 the problem is, this function is triggered before actual casting of skill, so removing this check caused autocast struct to be emptied, and skill failed to cast.

Reply:
@dastgirp This definitely needs to be adressed what 4144 mentioned.

Reply:
I just had a look at this and the condition should be changed to:
```c
if (sd != NULL && (ret == 0 || (skill_id != SA_ABRACADABRA && skill_id != WM_RANDOMIZESPELL)))
```